pkg/event/target/mqtt.go

```diff
@@ -134,19 +164,50 @@ func NewMQTTTarget(id string, args MQTTArgs) (*MQTTTarget, error) {
 		SetTLSConfig(&tls.Config{RootCAs: args.RootCAs}).
 		AddBroker(args.Broker.String())
 
+	var store Store
+
 	if args.QueueDir != "" {
-		options = options.SetStore(mqtt.NewFileStore(args.QueueDir))
+		store = NewQueueStore(args.QueueDir, 10000)
+		if oErr := store.Open(); oErr != nil {
+			return nil, oErr
+		}
+	} else {
+		store = NewMemoryStore(10000)
 	}
 
 	client := mqtt.NewClient(options)
+
+	// The client should establish a first time connection.
+	// Connect() should be successfull atleast once to publish events.
 	token := client.Connect()
-	if token.Wait() && token.Error() != nil {
-		return nil, token.Error()
-	}
 
-	return &MQTTTarget{
+	go func() {
+		// Repeat the pings until the client registers the clientId and recieves a token.
+		for {
+			if token.Wait() && token.Error() == nil {
+				fmt.Println("MQTT: Connected")
+				break
+			}
+			fmt.Println("MQTT: " + token.Error().Error() + ". Retrying in 5 seconds.")
+			time.Sleep(5 * time.Second)
+			token = client.Connect()
+		}
+	}()
+
+	target := &MQTTTarget{
 		id:     event.TargetID{ID: id, Name: "mqtt"},
 		args:   args,
 		client: client,
-	}, nil
+		store:  store,
+		reconn: false,
+	}
+
+	// Replay any previously persisted events in the store.
+	if len(target.store.ListAll()) != 0 {
+		go func() {
```

Comment:
Again `go target.retry()` perhaps is fine?

Reply:
Yes. Done.